azurelinuxagent/common/logcollector.py

```diff
@@ -0,0 +1,360 @@
+        LogCollector._log_to_results_file(header + output)
+
+    @staticmethod
+    def _parse_echo_command(message):
+        LogCollector._log_to_results_file(message)
+
+    @staticmethod
+    def _parse_copy_command(path):
+        file_paths = LogCollector._expand_path(path)
+        LogCollector._log_to_results_file(file_paths)
+        return file_paths
+
+    @staticmethod
+    def _convert_file_name_to_archive(file_name):
+        if file_name.startswith(TRUNCATED_FILES_DIR):
+            # /var/truncated/var/log/syslog.1 becomes truncated_var_log_syslog.1
+            original_file_path = file_name[len(TRUNCATED_FILES_DIR):].lstrip(os.path.sep)
+            archive_file_name = "truncated_" + original_file_path.replace(os.path.sep, "_")
+            return archive_file_name
+        else:
+            return file_name.lstrip(os.path.sep)
+
+    @staticmethod
+    def _convert_archive_name_to_file_name(archive_name):
+        truncated_prefix = "truncated_"
+        if archive_name.startswith(truncated_prefix):
+            # truncated_var_log_syslog.1 becomes /var/truncated/var/log/syslog.1
+            file_name = archive_name[len(truncated_prefix):].replace("_", os.path.sep)
+            original_file_name = os.path.join(TRUNCATED_FILES_DIR, file_name.lstrip(os.path.sep))
+            return original_file_name
+        else:
+            return os.path.join(os.path.sep, archive_name)
+
+    @staticmethod
+    def _remove_uncollected_truncated_files(files_to_collect):
+        # After log collection is completed, see if there are any old truncated files which were not collected
+        # and remove them since they probably won't be collected in the future. This is possible when the
+        # original file got deleted, so there is no need to keep its truncated version anymore.
+        truncated_files = os.listdir(TRUNCATED_FILES_DIR)
+
+        for file_path in truncated_files:
+            full_path = os.path.join(TRUNCATED_FILES_DIR, file_path)
+            if full_path not in files_to_collect:
+                rm_files(full_path)
+
+    @staticmethod
+    def _is_file_updated(file_name, archive_file):
+        # A file is updated if either its size or last modified time changed.
+        mtime_tarball = datetime.fromtimestamp(archive_file.mtime).replace(microsecond=0)
+        mtime_disk = datetime.fromtimestamp(os.path.getmtime(file_name)).replace(microsecond=0)
+
+        file_size_archive = archive_file.size  # uncompressed file size
+        file_size_disk = os.path.getsize(file_name)
+
+        return not (mtime_disk == mtime_tarball and file_size_disk == file_size_archive)
+
+    @staticmethod
+    def _get_list_of_files_in_archive():
+        with tarfile.open(OUTPUT_ARCHIVE_PATH, "a") as tarball:
+            return tarball.getnames()
+
+    @staticmethod
+    def _delete_file_from_archive(file):
+        try:
+            command_string = "tar --file {0} --delete {1}".format(OUTPUT_ARCHIVE_PATH, file)
```

Comment:
`tarfile` Python module doesn't have a method to delete a file from the archive, so command line option is used. Elsewhere, the Python API is preferred.

Reply:
Removed.